src/solvers/dg/2d/dg.jl

```diff
@@ -1849,47 +1907,37 @@ end
 # Calculate and store boundary flux across domain boundaries
 #NOTE: Do we need to dispatch on have_nonconservative_terms(dg.equations)?
 calc_boundary_flux!(dg::Dg2D, time) = calc_boundary_flux!(dg.elements.surface_flux_values,
-                                                          dg, time)
-function calc_boundary_flux!(surface_flux_values, dg::Dg2D, time)
+                                                          dg, time,
+                                                          dg.boundary_conditions...)
+function calc_boundary_flux!(surface_flux_values, dg::Dg2D, time, boundary_conditions...)
   @unpack surface_flux_function = dg
   @unpack u, neighbor_ids, neighbor_sides, node_coordinates, orientations = dg.boundaries
 
-  Threads.@threads for b in 1:dg.n_boundaries
-    # Fill outer boundary state
-    # FIXME: This should be replaced by a proper boundary condition
-    for i in 1:nnodes(dg)
-      @views u[3 - neighbor_sides[b], :, i, b] .= dg.initial_conditions(
-        node_coordinates[:, i, b], time, equations(dg))
-    end
-
-    # Get neighboring element
-    neighbor_id = neighbor_ids[b]
+  # Calculate indices
+  lasts = accumulate(+, dg.n_boundaries_per_direction)
+  firsts = lasts - dg.n_boundaries_per_direction + 1
 
-    # Determine boundary direction with respect to elements:
-    # orientation = 1: left -> 2, right -> 1
-    # orientation = 2: left -> 4, right -> 3
-    if orientations[b] == 1 # Boundary in x-direction
-      if neighbor_sides[b] == 1 # Element is on the left, boundary on the right
-        direction = 2
-      else # Element is on the right, boundary on the left
-        direction = 1
-      end
-    else # Boundary in y-direction
-      if neighbor_sides[b] == 1 # Element is below, boundary is above
-        direction = 4
-      else # Element is above, boundary is below
-        direction = 3
-      end
-    end
+  for direction in 1:4
+    Threads.@threads for b in firsts[direction]:lasts[direction]
+      # Get neighboring element
+      neighbor_id = neighbor_ids[b]
 
-    for i in 1:nnodes(dg)
-      # Call pointwise Riemann solver
-      u_ll, u_rr = get_surface_node_vars(u, dg, i, b)
-      flux = surface_flux_function(u_ll, u_rr, orientations[b], equations(dg))
+      for i in 1:nnodes(dg)
+        # Get boundary flux
+        u_ll, u_rr = get_surface_node_vars(u, dg, i, b)
+        if neighbor_sides[b] == 1 # Element is on the left, boundary on the right
+          u_inner = u_ll
+        else # Element is on the right, boundary on the left
+          u_inner = u_rr
+        end
+        x = get_node_coords(node_coordinates, dg, i, b)
+        flux = boundary_conditions[direction](u_inner, orientations[b], direction, x, time,
+                                              surface_flux_function, equations(dg))
```

Comment:
That's not type-stable. I guess a lot of the allocations and the additional overhead come from this part.

Reply:
Are you sure (you probably are)? I thought (by intuition) that by splatting `dg.boundary_conditions` and slurping it again, `boundary_conditions` in this function is a `Tuple` of concrete types and thus `boundary_conditions[direction]` should be type stable... but I guess not. Any suggestion for a quick fix?

Otherwise, I'd suggest to keep it as is, add a strongly-worded comment, and revisit this once Taal is alive.

Reply:
The iteration `for direction in 1:4` isn't unrolled, so `boundary_conditions[direction]` isn't known.
```julia
julia> function foo(args...)
           res = 0
           for i in 1:4
               res += args[i](1)
           end
           res
       end
foo (generic function with 1 method)

julia> @code_warntype foo(+, -, sin, cos)
Variables
  #self#::Core.Compiler.Const(foo, false)
  args::Core.Compiler.Const((+, -, sin, cos), false)
  res::Any
  @_4::Union{Nothing, Tuple{Int64,Int64}}
  i::Int64

Body::Any
1 ─       (res = 0)
│   %2  = (1:4)::Core.Compiler.Const(1:4, false)
│         (@_4 = Base.iterate(%2))
│   %4  = (@_4::Core.Compiler.Const((1, 1), false) === nothing)::Core.Compiler.Const(false, false)
│   %5  = Base.not_int(%4)::Core.Compiler.Const(true, false)
└──       goto #4 if not %5
2 ┄ %7  = @_4::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│         (i = Core.getfield(%7, 1))
│   %9  = Core.getfield(%7, 2)::Int64
│   %10 = res::Any
│   %11 = Base.getindex(args, i)::Any
│   %12 = (%11)(1)::Any
│         (res = %10 + %12)
│         (@_4 = Base.iterate(%2, %9))
│   %15 = (@_4 === nothing)::Bool
│   %16 = Base.not_int(%15)::Bool
└──       goto #4 if not %16
3 ─       goto #2
4 ┄       return res
```

Reply:
The line `│   %11 = Base.getindex(args, i)::Any` demonstrates that point.

Reply:
If you want to make that part type stable, you need to use something like [Unrolled.jl](https://github.com/cstjean/Unrolled.jl):
```julia
julia> using Unrolled

julia> @unroll function bar(args)
           res = 0
           @unroll for idx in 1:length(args)
               res += args[idx](1)
           end
           res
       end
bar_unrolled_expansion_ (generic function with 1 method)

julia> @code_warntype bar((+, -, sin, cos))
Variables
  #self#::Core.Compiler.Const(bar, false)
  args::Core.Compiler.Const((+, -, sin, cos), false)
  res::Union{Float64, Int64}
  idx@_4::Int64
  idx@_5::Int64
  idx@_6::Int64
  idx@_7::Int64

Body::Float64
1 ─       (res = 0)
│         (idx@_4 = 1)
│   %3  = res::Core.Compiler.Const(0, false)::Core.Compiler.Const(0, false)
│   %4  = Base.getindex(args, idx@_4::Core.Compiler.Const(1, false))::Core.Compiler.Const(+, false)
│   %5  = (%4)(1)::Core.Compiler.Const(1, false)
│         (res = %3 + %5)
│         (idx@_5 = 2)
│   %8  = res::Core.Compiler.Const(1, false)::Core.Compiler.Const(1, false)
│   %9  = Base.getindex(args, idx@_5::Core.Compiler.Const(2, false))::Core.Compiler.Const(-, false)
│   %10 = (%9)(1)::Core.Compiler.Const(-1, false)
│         (res = %8 + %10)
│         (idx@_6 = 3)
│   %13 = res::Core.Compiler.Const(0, false)::Core.Compiler.Const(0, false)
│   %14 = Base.getindex(args, idx@_6::Core.Compiler.Const(3, false))::Core.Compiler.Const(sin, false)
│   %15 = (%14)(1)::Float64
│         (res = %13 + %15)
│         (idx@_7 = 4)
│   %18 = res::Float64::Float64
│   %19 = Base.getindex(args, idx@_7::Core.Compiler.Const(4, false))::Core.Compiler.Const(cos, false)
│   %20 = (%19)(1)::Float64
│         (res = %18 + %20)
│   %22 = res::Float64::Float64
└──       return %22
```

Reply:
Thank you very much for these illustrations!